src/plugins/projectexplorer/projectvcsstatus.cpp

```diff
@@ -15,18 +15,20 @@ namespace ProjectExplorer {
 
 namespace {
 
-QSet<FolderNode*> extractFolderNodes(const QHash<Utils::FilePath, Core::VcsChangeType> &fileChanges, const Utils::FilePath& topLevel)
+QSet<FolderNode*> extractFolderNodes(const QHash<Utils::FilePath, Core::VcsChangeType> &fileChanges, Project *project)
 {
     QSet<FolderNode*> result;
-    auto *topNode = ProjectTree::nodeForFile(topLevel);
-    if (topNode == nullptr) {
+    if (fileChanges.count() < 1 || project->containerNode() == nullptr) {
         return result;
     }
-    auto *topFolder = topNode->asFolderNode();
-    if (topFolder == nullptr) {
+    result.insert(project->containerNode());
+
+    auto *projectNode = project->rootProjectNode();
+    if (projectNode == nullptr || projectNode->asFolderNode() == nullptr) {
         return result;
     }
-    result.insert(topFolder);
+    result.insert(projectNode->asFolderNode());
```

Comment:
Something bugs me about all these conditions. Can we structure them in a more readable way, that avoids combining and repeating stuff.

Reply:
I understand what you mean but `project->containerNode()`, `project->rootProjectNode()` and `project->rootProjectNode()->asFolderNode()` was never checked. So these check must be executed at least once. We can put the upper section of this function in a lambda to make it more expressive that this code is responsible for adding root nodes. But this also introduces a check for the return of this new lambda.